### PR TITLE
Multivariate marlin

### PR DESCRIPTION
**Pull request by ryanleh**

### CODE REVIEW

**Comment by weikengchen** [comment]
I reconcile this PR with the upstream, and I think it is ready to review and merge.

This one might need prioritization since it changes and relocates Marlin. Other PRs easily conflict with it.

File: src/marlin/marlin_pst13_pc/combinations.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
+//! Compute all combinations of values in a given list
+//! Credit: https://github.com/meltinglava/uniquecombinations/
```
Comment (weikengchen) [inline src/marlin/marlin_pst13_pc/combinations.rs:2]: I checked. This repo is under Apache 2 and MIT license, so it fits well into our license. Any additional reference should we put?

**Review by weikengchen — APPROVED** [review]
I went through the files. I think it is in a good shape.

**Comment by weikengchen** [comment]
Let me @Pratyush here.

This one may need some prioritized review since it is likely to conflict a lot with the constraints PR. I feel it is better to merge this one soon, and then the constraints PR can adjust correspondingly.

**Comment by weikengchen** [comment]
_Let me ping Pratyush again in case he forgot :)_

**Comment by weikengchen** [comment]
@Pratyush

File: src/marlin/marlin_pc/mod.rs [hunk 2]
```diff
@@ -232,11 +124,8 @@ where
                 if enforced_degree_bounds.is_empty() {
                     (None, None)
                 } else {
-                    let mut sorted_enforced_degree_bounds = enforced_degree_bounds.clone();
-                    sorted_enforced_degree_bounds.sort();
-
```
Comment (Pratyush) [inline src/marlin/marlin_pc/mod.rs:237]: Why was this removed? I think we need it, otherwise the `lowest_shifted_power` value will be incorrect.

File: src/marlin/marlin_pc/mod.rs [hunk 3]
```diff
@@ -390,8 +279,6 @@ where
             let challenge_j = opening_challenges(opening_challenge_counter);
             opening_challenge_counter += 1;
 
-            assert_eq!(degree_bound.is_some(), rand.shifted_rand.is_some());
```
Comment (Pratyush) [inline src/marlin/marlin_pc/mod.rs:393]: Why was this assertion removed?

File: src/marlin/marlin_pst13_pc/data_structures.rs [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+    /// prepared for use in pairings
+    #[derivative(Debug = "ignore")]
+    pub prepared_beta_h: Vec<E::G2Prepared>,
+    /// The number of variables `self` is initialized for
+    pub num_vars: usize,
+    /// The maximum degree supported by `self`
+    pub max_degree: usize,
+}
+
+impl<E, P> CanonicalSerialize for UniversalParams<E, P>
```
Comment (Pratyush) [inline src/marlin/marlin_pst13_pc/data_structures.rs:51]: Is there a reason the `derive` doesn't work here?

Comment (weikengchen) [reply src/marlin/marlin_pst13_pc/data_structures.rs:51]: Due to `G2Prepared`. Ideally, we need `#[derivative(CanonicalSerialize = "ignore")]` so some of the entries are ignored during serialization. Nevertheless, for such, we still need to implement CanonicalDeserialize.

File: src/marlin/marlin_pc/mod.rs (hunk 2 above)
Comment (weikengchen) [reply src/marlin/marlin_pc/mod.rs:237]: This might be due to incorrect sync with the master branch. Let me take a look

File: src/marlin/marlin_pc/mod.rs (hunk 3 above)
Comment (weikengchen) [reply src/marlin/marlin_pc/mod.rs:393]: likely also due to incorrect sync, let me also take a look

**Comment by weikengchen** [comment]
Seems the synchronization with the master has brought some unexpected changes, I will take a look more carefully.

**Comment by weikengchen** [comment]
(I am back from a deadline and will be wrapping up this)
